modules/data.land/R/soilgrids_texture_extract.R

```diff
@@ -0,0 +1,129 @@
+    lonlat <-cbind(internal_site_info$site_info.lon, internal_site_info$site_info.lat)
+    depths <-c("0-5cm", "5-15cm", "15-30cm", "30-60cm", "60-100cm", "100-200cm")
+    
+    # Reproject locations to soilgrids projection
+    # Soilgrids data is using Homolosine projection https://www.isric.org/explore/soilgrids/faq-soilgrids
+    p <-terra::vect(lonlat, crs = "+proj=longlat +datum=WGS84") # Users need to provide lon/lat
+    newcrs <- "+proj=igh +datum=WGS84 +no_defs +towgs84=0,0,0"
+    p_reproj <- terra::project(p, newcrs) # Transform the point vector to data with Homolosine projection
+    data_tag <- c("_mean.vrt", "_Q0.05.vrt", "_Q0.5.vrt", "_Q0.95.vrt")
+    name_tag <- expand.grid(depths, data_tag, stringsAsFactors = F) #find the combinations between data and depth tags.
+    L <- split(as.data.frame(name_tag), seq(nrow(as.data.frame(name_tag))))#convert tags into lists.
+    
+    if ("try-error" %in% class(try(soilt_real <- L %>% furrr::future_map(function(l){
+      soilt_url <- paste0(base_data_url, l[[1]], l[[2]])
+      soilt_map <- terra::extract(terra::rast(soilt_url), p_reproj)
+      unlist(soilt_map[, -1])/10
+    }, .progress = T)))) {
+      soilt_real <- vector("list", length = length(L))
+      pb <- utils::txtProgressBar(min = 0, max = length(L), style = 3)
+      for (i in seq_along(L)) {
+        l <- L[[i]]
+        soilt_url <- paste0(base_data_url, l[[1]], l[[2]])
+        soilt_map <- terra::extract(terra::rast(soilt_url), p_reproj)
+        soilt_real[[i]] <- unlist(soilt_map[, -1])/10
+        utils::setTxtProgressBar(pb, i)
+      }
+    }
+    
+    for (dep in seq_along(depths)) {
+      dep.ind <- which(grepl(depths[dep], name_tag[, 1]))
+      soiltquant[dep, ] <- soilt_real[dep.ind] %>% unlist
+    }
+    
+    
+    # Parse extracted data and prepare for output
+    quantile_name <-c(paste("Mean_", site_info$site_id, sep = ""),paste("0.05_", site_info$site_id, sep = ""),paste("0.5_", site_info$site_id, sep = ""),paste("0.95_", site_info$site_id, sep = ""))
+    colnames(soiltquant) <- quantile_name
+    soilt_dep <- cbind(soiltquant, depths)
+    soilt_df <- tidyr::pivot_longer(as.data.frame(soilt_dep),cols = tidyselect::all_of(quantile_name),names_to = c("Quantile", "Siteid"),names_sep = "_")
+    # Remove NA
+    soilt_df <- stats::na.omit(soilt_df)
+    colnames(soilt_df) <- c("Depth", "Quantile", "Siteid", "Value")
+    soilt_df$Value<-as.numeric(soilt_df$Value)
+    
+    if (!is.null(outdir)) {
+      PEcAn.logger::logger.info(paste0("Storing results in: ",outdir))
+      saveRDS(soilt_df,file=outdir)
+    }
+    else {
+      PEcAn.logger::logger.error("No output directory found.")
+    }
+    # Return the results to the terminal as well
+    return(soilt_df)
+  }
+  
+  
+  if (future::supportsMulticore()) {
+    future::supportsMulticore()
+    future::plan(future::multicore)
+   } else {
+    future::plan(future::multisession,workers=3)
+  }
+  
+  sand_data_url <-
```

Comment:
In a real situation where we might already have the `SoilGrids` data downloaded on the server, we might want to replace the `VRT` URL with the actual physical path to the data. To do this, we might need to set those VRT paths as the function arguments (e.g., `list(sand_data_url = "/vsicurl?max_retry=30&retry_delay=60&list_dir=no&url=https://files.isric.org/soilgrids/latest/data/sand/sand_")`). The downloaded file path, for example on the SCC, can be `/projectnb/dietzelab/dongchen/anchorSites/NA_runs/soil_nc/soilgrids_250m/sand/sand_0-5cm_mean/sand/sand_0-5cm_mean.tif`.